daemon/network.go

```diff
@@ -99,15 +100,45 @@ func (daemon *Daemon) getAllNetworks() []libnetwork.Network {
 	return daemon.netController.Networks()
 }
 
-func isIngressNetwork(name string) bool {
-	return name == "ingress"
+type ingressJob struct {
+	create *clustertypes.NetworkCreateRequest
+	ip     net.IP
 }
 
-var ingressChan = make(chan struct{}, 1)
+var (
+	ingressWorkerOnce  sync.Once
+	ingressJobsChannel chan *ingressJob
+	ingressWorkerStop  chan struct{}
+	ingressID          string
+)
+
+func (daemon *Daemon) setupIngressWorker() {
+	ingressWorkerStop = make(chan struct{}, 1)
+	ingressJobsChannel = make(chan *ingressJob, 100)
+	go func() {
+		for {
+			select {
+			case <-ingressWorkerStop:
+				ingressID = ""
+				return
+			case r := <-ingressJobsChannel:
+				if r.create != nil {
+					daemon.setupIngress(r.create, r.ip, ingressID)
+					ingressID = r.create.ID
+				} else {
+					daemon.releaseIngress(ingressID)
+					ingressID = ""
+				}
+			}
+		}
+	}()
+}
 
-func ingressWait() func() {
-	ingressChan <- struct{}{}
-	return func() { <-ingressChan }
+func (daemon *Daemon) stopIngressWorker() {
+	if ingressWorkerStop != nil {
```

Comment:
This `if` statement also strikes me as suspicious. It implies that `stopIngressWorker` can be called before `setupIngressWorker`. But if that's the case, it suggests that after leaving a swarm and rejoining, we could close an already-closed channel (`ingressWorkerStop` never gets reset to nil).

Again, I think adding locking around this stuff would make it a lot easier to be confident that there aren't concurrency bugs. If nothing else, it would make my review a lot easier.